### ISSUE TITLE
getting errors in training and while inference the model

### ISSUE THREAD

**Issue by uttu-parashar** [PattanaikL/GeoMol#4]
1. I have created a new environment using your .sh file and running the training script with the same datasets. But I am getting this error.
`RuntimeError: Sizes of tensors must match except in dimension 0. Expected size 21 but got size 19 for tensor number 1 in the list.`

2. While running your generate_confs script I am getting this error. in this line ( data = Batch.from_data_list(data_list=[tg_data]) )
`TypeError: argument of type 'int' is not iterable`

3. if I am passing data directly to the model bypassing above line then in line 
   `model(tg_data, inference=True, n_model_confs=n_confs*2`  I am getting this error 
   `AttributeError: 'GlobalStorage' object has no attribute 'bincount'` 
   NOTE : While passing the data directly to the model i changed ` n_atoms_per_mol = data.batch.bincount()` TO
    `n_atoms_per_mol = data.bincount()`  in get_neighbor_ids function of model.utils script. If i am not changing this line then the error  
   is like `NoneType attribute has no attribute bincount()`

![Screenshot from 2021-12-31 15-24-12](https://user-images.githubusercontent.com/61959483/147817086-e7a2faf2-51c9-41c4-99d7-dbb0eeb1c0ce.png)
![Screenshot from 2021-12-31 15-24-17](https://user-images.githubusercontent.com/61959483/147817087-d8cb845e-02ca-4e94-a6cb-e9194cf556a5.png)
![Screenshot from 2021-12-31 15-24-33](https://user-images.githubusercontent.com/61959483/147817088-45e6bc67-d365-4a74-8a35-2d5f1bc80199.png)

**Comment by PattanaikL** [comment]
This seems related to #2. Which version of torch-geometric and pytorch are you using? Could you try downgrading to torch-geometric 1.6.3. and pytorch to 1.7.0 to see if this helps?

**Comment by uttu-parashar** [comment]
thanks for the reply... yes, it is working with older versions of torh_geometric and torch. but I want to make it work on the latest versions of torch_geometric and torch. any suggestion you want to suggest.? like where I should make changes to make it work on torch and torch_geometric latest versions.?

**Comment by forechoandlook** [comment]
> thanks for the reply... yes, it is working with older versions of torh_geometric and torch. but I want to make it work on the latest versions of torch_geometric and torch. any suggestion you want to suggest.? like where I should make changes to make it work on torch and torch_geometric latest versions.?

Well, I get the same problem that `data.x` has different shape and can not be concated together.  It uses `one-hot` encoding and different molecular has different atom counts.   

![image](https://user-images.githubusercontent.com/50866639/149051253-3d231960-7a19-425e-8603-d21bb40044e7.png)

**Comment by uttu-parashar** [comment]
yes, when DataLoader creates batches of data it concatenates the tensor of values internally. That's wherein the new version of torch_geometric and torch shape mismatch error is occurring.

**Comment by uttu-parashar** [comment]
`data.x` is the first attribute where this error is occurring. I  think the same error might occur when DataLoader will try to concatenate other attributes like eddge_attr to create batches.

**Comment by PattanaikL** [comment]
Hi, just wanted to come back to this. I haven't had a chance to play around with the new version of torch_geometric, which I suspect is causing most of the issues. Were you able to get the code to work with pyg 2.0? If not, would this be helpful? I'm considering putting in some effort to upgrade the repo to pyg 2.0 if it's useful for folks

**Comment by uttu-parashar** [comment]
The latest version of torch_geometric this repo is working with is 1.7.2. After 1.7.2 they directly released torch_geometric 2.0 versions and some major changes might be there. The problem is occurring in torch_geometric 2.0 versions when DataLoader is trying to create baches of data objects. For creating batches it needs to concatenate data object attributes in the required right dimensions and that's where some attributes are being concatenated in wrong dimensions and shape mismatch error is occurring. I tried writing my collate function but torch_geometric is deleting our collate function internally as soon as we are passing it. I also tried replacing their DataLoader with torch DataLoader with Our collate function -- ``` def collate_fn(batchh):
    my_batch = torch_geometric.data.batch.Batch()
    batch_obj = []
    for i in range(len(batchh)):
        batch_obj = batch_obj+[i]*batchh[i].x.size()[0]
    my_batch.batch = torch.tensor(batch_obj)
    my_batch.boltzmann_weight = torch.tensor([obj.boltzmann_weight for obj in batchh])
    my_batch.chiral_tag = torch.cat([obj.chiral_tag for obj in batchh],dim=0)
    my_batch.degeneracy = torch.tensor([obj.degeneracy for obj in batchh])
    my_batch.edge_attr = torch.cat([obj.edge_attr for obj in batchh],dim=0)
    my_batch.edge_index = torch.cat([obj.edge_index for obj in batchh],dim=1)
    my_batch.edge_index_dihedral_pairs = torch.cat([obj.edge_index_dihedral_pairs for obj in batchh],dim=1)
    my_batch.mol = [obj.mol for obj in batchh]
    my_batch.name = [obj.name for obj in batchh]
    my_batch.neighbors = [obj.neighbors for obj in batchh]
    my_batch.pos = [obj.pos for obj in batchh]
    my_batch.pos_mask = torch.cat([obj.pos_mask for obj in batchh],dim=0)
    my_batch.x = torch.cat([obj.x for obj in batchh],dim=0)
    my_batch.z = torch.cat([obj.z for obj in batchh],dim=0)

    return my_batch ``` 
    
    with this experiment I was able to create batches of data objects but in the model class  ```         
   for i, (s, e) in enumerate(self.dihedral_pairs.t()):
            # this indicates which neighbor is the correct x <--> y map (see overleaf doc)
            self.x_map_to_neighbor_y[i, 0:len(self.neighbors[s.item()])] = self.neighbors[s.item()] == e
            self.y_map_to_neighbor_x[i, 0:len(self.neighbors[e.item()])] = self.neighbors[e.item()] == s ``` 
    here i was getting keyerrors in ``` def assign_neighborhoods(self, x, edge_index, edge_attr, batch, data): ``` Function

**Comment by gcorso** [comment]
Any updates on this issue?

**Comment by octavian-ganea** [comment]
did you try this code https://github.com/PattanaikL/GeoMol/blob/main/count_geomol_failures.ipynb ? It shows exactly which packages you need.

**Comment by kiranfranklin999** [comment]
any update on this issue.?
